hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/DumpS3GuardTable.java

```diff
@@ -0,0 +1,691 @@
+   * @throws Exception failure.
+   */
+  @Override
+  protected void serviceStart() throws Exception {
+    if (getStore() == null) {
+      List<String> arg = getArgumentList(2, 2, USAGE_MESSAGE);
+      bindFromCLI(arg.get(0));
+      destPath = arg.get(1);
+    }
+  }
+
+  /**
+   * Dump the filesystem and the metastore.
+   * @return the exit code.
+   * @throws ServiceLaunchException on failure.
+   * @throws IOException IO failure.
+   */
+  @Override
+  public int execute() throws ServiceLaunchException, IOException {
+
+    try {
+      final File scanFile = new File(
+          destPath + SCAN_CSV).getCanonicalFile();
+      scanFile.getParentFile().mkdirs();
+
+      try (CsvFile csv = new CsvFile(scanFile);
+           DurationInfo ignored = new DurationInfo(LOG,
+               "scanFile dump to %s", scanFile)) {
+        scanMetastore(csv);
+      }
+
+      if (getFilesystem() != null) {
+
+        Path basePath = getFilesystem().qualify(new Path(getUri()));
+
+        final File destFile = new File(destPath + STORE_CSV)
+            .getCanonicalFile();
+        LOG.info("Writing Store details to {}", destFile);
+        try (CsvFile csv = new CsvFile(destFile);
+             DurationInfo ignored = new DurationInfo(LOG, "List metastore")) {
+
+          LOG.info("Base path: {}", basePath);
+          dumpMetastore(csv, basePath);
+        }
+
+        // these operations all update the metastore as they list,
+        // that is: they are side-effecting.
+        final File treewalkFile = new File(destPath + TREE_CSV)
+            .getCanonicalFile();
+
+        try (CsvFile csv = new CsvFile(treewalkFile);
+             DurationInfo ignored = new DurationInfo(LOG,
+                 "Treewalk to %s", treewalkFile)) {
+          treewalkFilesystem(csv, basePath);
+        }
+        final File flatlistFile = new File(
+            destPath + FLAT_CSV).getCanonicalFile();
+
+        try (CsvFile csv = new CsvFile(flatlistFile);
+             DurationInfo ignored = new DurationInfo(LOG,
+                 "Flat list to %s", flatlistFile)) {
+          listStatusFilesystem(csv, basePath);
+        }
+        final File rawFile = new File(
+            destPath + RAW_CSV).getCanonicalFile();
+
+        try (CsvFile csv = new CsvFile(rawFile);
+             DurationInfo ignored = new DurationInfo(LOG,
+                 "Raw dump to %s", rawFile)) {
+          dumpRawS3ObjectStore(csv);
+        }
+        final File scanFile2 = new File(
+            destPath + SCAN2_CSV).getCanonicalFile();
+
+        try (CsvFile csv = new CsvFile(scanFile);
+             DurationInfo ignored = new DurationInfo(LOG,
+                 "scanFile dump to %s", scanFile2)) {
+          scanMetastore(csv);
+        }
+
+      }
+
+      return LauncherExitCodes.EXIT_SUCCESS;
+    } catch (IOException | RuntimeException e) {
+      LOG.error("failure", e);
+      throw e;
+    }
+  }
+
+  /**
+   * Dump the filesystem via a recursive treewalk.
```

Comment:
In my fsck patch I'm using iterative walk on the tree. Why? Deep fs structure can build up huge stacks, I tend to generally avoid using recursive walk because of that. 
What's against the iterative approach here?

Reply:
What do you mean? Build a queue of paths to scan? I was just doing "traditional treewalk strategy", nothing else

Reply:
I prefer the queue, but this could a be a good solution as well.